atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractLockAwareTransactionManager.java

```diff
@@ -51,66 +37,16 @@
-            HeldLocksToken lockToken = null;
-            if (lockRequest != null) {
-                Validate.isTrue(lockRequest.getVersionId() == null, "Using a version id is not allowed");
-                HeldLocksToken response = getLockService()
-                        .lockAndGetHeldLocks(LockClient.ANONYMOUS.getClientId(), lockRequest);
-                if (response == null) {
-                    RuntimeException ex = new LockAcquisitionException(
-                            "Failed to lock using the provided lock request: " + lockRequest);
-                    log.warn("[{}] Could not lock successfully", runId, ex);
-                    failureCount++;
-                    if (shouldStopRetrying(failureCount)) {
-                        log.warn("[{}] Failing after {} tries", runId, failureCount, ex);
-                        throw ex;
-                    }
-                    sleepForBackoff(failureCount);
-                    continue;
-                }
-                lockToken = response;
-            }
-
+        checkOpen();
+        Supplier<AdvisoryLocksCondition> conditionSupplier =
+                AdvisoryLockConditionSuppliers.get(getLockService(), lockTokens, lockSupplier);
+        return runTaskWithConditionWithRetry(conditionSupplier, (transaction, condition) -> {
             try {
-                if (lockToken == null) {
-                    T result = runTaskWithLocksThrowOnConflict(lockTokens, task);
-                    logSuccess(runId, failureCount);
-                    return result;
-                } else {
-                    T result = runTaskWithLocksThrowOnConflict(IterableUtils.append(lockTokens, lockToken), task);
-                    logSuccess(runId, failureCount);
-                    return result;
-                }
-            } catch (TransactionFailedException e) {
-                if (!e.canTransactionBeRetried()) {
-                    log.warn("[{}] Non-retriable exception while processing transaction.", runId, e);
-                    throw e;
-                }
-                if (e instanceof TransactionLockTimeoutException) {
-                    refreshAfterLockTimeout(lockTokens, (TransactionLockTimeoutException) e);
-                }
-                failureCount++;
-                if (shouldStopRetrying(failureCount)) {
-                    log.warn("[{}] Failing after {} tries", runId, failureCount, e);
-                    throw e;
-                }
-                log.info("[{}] Retrying transaction", runId, e);
-            } catch (RuntimeException e) {
-                log.warn("[{}] RuntimeException while processing transaction.", runId, e);
-                throw e;
+                return task.execute(transaction, condition.getLocks());
             } finally {
-                if (lockToken != null) {
-                    getLockService().unlock(lockToken.getLockRefreshToken());
-                }
+                condition.cleanup();
```

Comment:
finally block shouldn't be needed here, as runTaskWithConditionWithRetry should presumably also include it - there's no special logic here for the lock case.